check.js

```diff
@@ -0,0 +1,459 @@
+    return res;
+  }
+
+  res.result = true
+  return res;
+}
+
+/**
+ * Check if an address can set a spawn proxy for the given ship.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} address - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canSetSpawnProxy(contracts, prefix, address) {
+  let res = { result: false };
+  let parentShip = await ships.getShip(contracts, prefix);
+  // must be the owner of the ship
+  if (!await ships.isOwner(contracts, parentShip, address)) {
+    res.reason = reasons.permission;
+    return res;
+  }
+  // the ship must be active
+  if (!await ships.isActive(contracts, parentShip)) {
+    res.reason = reasons.inactive;
+    return res;
+  }
+  res.result = true;
+  return res;
+}
+
+/**
+ * Check if the sender address can send the provided ship to the target
+ * address.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} sender - Sender's address.
+ * @param {String} target - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canTransferShip(contracts, ship, source, target) {
+  let res     = { result: false };
+  // prevent burning of ships
+  if (utils.isZeroAddress(target))
+  {
+    res.reason = reasons.zero;
+    return res;
+  }
+
+  let shipObj = await ships.getShip(contracts, ship);
+  // must be either the owner of the ship,
+  // a transfer proxy for the ship,
+  // or an operator for the owner
+  if (!ships.isOwner(contracts, shipObj, source) &&
+      !ships.isTransferProxy(contracts, shipObj, source) &&
+      !await ships.isOperator(contracts, shipObj.owner, source))
+  {
+    res.reason = reasons.permission;
+    return res;
+  }
+
+  res.result = true;
+  return res;
+}
+
+/**
+ * Check if the address can set a transfer proxy for the ship.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} address - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canSetTransferProxy(contracts, ship, address) {
+  let res     = { result: false };
+  let shipObj = await ships.getShip(contracts, ship);
+  // must be either the owner of the ship,
+  // or an operator for the owner
+  if (!ships.isOwner(contracts, shipObj, address) &&
+      !await ships.isOperator(contracts, shipObj.owner, address))
+  {
+    res.reason = reasons.permission;
+    return res;
+  }
+  res.result = true;
+  return res;
+}
+
+/**
+ * Check if the address can configure public keys for the ship.
+ * @param {Object} contracts - An Urbit contracts object.
+ * @param {Number} tokenId - Ship token.
+ * @param {String} address - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canConfigureKeys(contracts, ship, address) {
+  res = { result: false };
```

Comment:
Does this work without `let`?

Reply:
It.. apparently does? 😕  That function gets called a couple of times in the test suite, and the tests are passing.  I'll fix it, though.